Refactors of some developer/contributor flows (#440)
demisto/content-docs: docs/contributing/contributing.md

title: Contributing 
---

Thanks for being interested in contributing to Cortex XSOAR. This document describes the Contribution process. If you are not sure whether you should read this, make sure you read the [Getting Started Guide](../concepts/getting-started-guide) first.

Contributing allows you to make the content that you build on Cortex XSOAR available to every client through the  [Marketplace](../partners/marketplace). Content can be either Partner or Community supported, [Free or Paid](../partners/marketplace#pricing).

All the free content (i.e. everything excluding Paid Content Packs) is open source and lives in the Cortex XSOAR [GitHub Repository](https://github.com/demisto/content), with a MIT license.

## Support and requirements

> Contributions can be either officially supported (by Palo Alto Networks, a Technology Partner or a developer), or *community* supported: the former means that when a customer has a problem with the content, they will have an email address or web site to reach out to, and they will expect an answer. When the contribution is *community* supported, it's not required to provide support (customers will ask questions on our [Live Community Forum](https://live.paloaltonetworks.com/t5/cortex-xsoar-discussions/bd-p/Cortex_XSOAR_Discussions)) and the quality requirements for the contribution to be accepted will be lower.

Author: done

For more information about the different support levels, check the [official documentation](https://docs.paloaltonetworks.com/cortex/cortex-xsoar/6-0/cortex-xsoar-admin/marketplace/marketplace-overview/content-packs-support-types.html).

## How to contribute

After you have created your content, you must submit your content to Palo Alto Networks: the Cortex XSOAR Content Team will review and approve it before it becomes available to customers.

There are three ways to submit your work:
 1. Contribute from the Cortex XSOAR UI [Marketplace](../contributing/marketplace). This flow is simpler and doesn't require to set up a development environment or be familiar with git. However, it is recommended **only** in the following scenarios:
    - You are an individual contributor, contributing **new** content that is *community* supported.
    - You are making small changes to **existing** content, even if it's *xsoar* or *partner* supported (usually a bug fix or adding a new command to an Integration).
 1. Contribute through a GitHub Pull Request on the public [XSOAR Content Repository](https://github.com/demisto/content). Use this flow in the following scenarios:

>      - You are a Technology Partner or a developer contributing with *partner* supported new content.

Author: done

     - Your contribution is big and contains lots of different parts (Integrations, Scripts, Playbooks, Layouts, etc) that is likely to lead to a complex review process.
     - You are proficient with GitHub.
 1. Contribute through a private GitHub repository: this is required if you are providing a **Premium** (aka Paid) Content Pack. The requirements are the same of supported Packs, and the contribution process is described [here](../integrations/premium_packs).

> This document describes the main flow that covers *supported* contributions and summarizes everything you must do before and after opening a Pull Request on GitHub to contribute your pack.

Author: done

If you are contributing *community* supported content, feel free to skip the rest of this document (although it's still recommended to read it to be aware of the best practices).

## Contributor Guidelines

Please read the following guidelines carefully: following them will maximize the chances for a fast, easy, and effective review process for everyone involved. If something is not clear, please don't hesitate to reach out to us via [Slack](http://go.demisto.com/join-our-slack-community) on the `#demisto-developers` channel.

1. Begin by designing your contribution: we recommend to follow the [Design](../concepts/design) guidelines to identify what you want to build and make sure it is aligned with our best practices. Also check out the [Design Tutorial](../tutorials/tut-design).
1. Make sure you have all the [Contributing Requirements](../contributing/contrib-requirements) satisfied.
1. Setup a development environment by following the brief [Dev Setup Guide](dev-setup) or the more detailed [Tutorial](../tutorials/tut-setup-dev).

> 1. Review the [Contribution Checklist](../contributing/checklist).

Reviewer: Let's push this to the end

Author: done

> 1. Follow the [Content Pack format](packs-format) to build your contribution. [demisto-sdk init](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/init/README.md) will help you create it.

Reviewer: ```suggestion
1. Follow the [Content Pack Structure](../docs/packs/packs-format) to build your contribution. [demisto-sdk init](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/init/README.md) will help you create it.
```

Author: done

1. Depending on the content entities you need to build, navigate to the specific section of this website for details. If you are creating Integrations and/or Automations, make sure that you:

>     * Use the proper  [Directory Structure](../integrations/package-dir). [demisto-sdk init](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/init/README.md) will help you create it. If working on existing code, beyond trivial changes, we require converting to this structure as it allows running linting and unit tests and provides a clearer review process.

Author: Not sure if I understand what you mean : this statement is true for both Integrations and Automations, so "Directory Structure" should be generic, not just "Integration Directory Structure". and "Integration and Automations Directory Structure" is too long IMO.

>     * Use the proper  [Directory Structure](../integrations/package-dir). [demisto-sdk init](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/init/README.md) will help you create it. If working on existing code, beyond trivial changes, we require converting to this structure as it allows running linting and unit tests and provides a clearer review process.

Reviewer: There are two things mentioned here:
1. The hyperlink text (Directory Structure) is too generic (especially since we have the pack structure which is a bit similar), so I wanted to change it a bit.
2. The article itself has a confusing title, again - Directory Structure. Then U suggested the other one. I agree it's a bit long, but I want to differentiate this article from the pack structure one, after all, it explains only the integration/automation directory structure.

    * Understand the [YAML file](../integrations/yaml-file) structure and the [Parameter Types](../integrations/parameter-types).
    * Make sure your integration follows our [Logo Guidelines](../integrations/integration-logo).
    * Read and follow [Python code conventions](../integrations/code-conventions) (recommended) or [Powershell code conventions](../integrations/powershell-code) (advanced users only).
    * If your integration generates Incidents, follow the [Fetch Incidents](../fetching-incidents) guidelines.
    * Make sure your commands make proper use of the [Context](../integrations/context-and-outputs), including [Context Standards](../integrations/context-standards-about) and [DBotScore](../integrations/dbot).
    * Run and verify that the various linters we support pass as detailed [here](../integrations/linting).
    * Make sure to create unit tests as documented [here](../integrations/unit-testing)

>     * Document your integration and automation as detailed [here](../documentation/readme_file) and [here](../documentation/documentation_tips).

Reviewer: ```suggestion
    * Document your integration and automation as detailed in the [README guide](../documentation/readme_file) and in the [Documentation Best Practices](../documentation/documentation_tips) article.
```

Author: done

> 1. Make sure your Content Pack is properly [documented](../documentation/pack-docs) and read the [documentation best practices](../documentation/documentation_tips).

Reviewer: The documentation_tips is mentioned twice. Where do you think is better?

Author: Added a line with the best practices and remove the other two. done.

> 1. Validate your content: the validation hook should run automatically every time you `git commit`. You can also run the validation manually by using [demisto-sdk validate](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/validate/README.md). 

Reviewer: This also appears in the `Before opening a Pull Request`

Author: done (left only one)

1. As you build newer versions of your Content Pack, document your changes in a relevant release notes file as detailed [here](../documentation/release-notes).

> At this point you should be ready to submit a Pull Request! Check out again our [Contributing Checklist](../contributing/checklist), and for more details on the review process, refer to our [PR Conventions](../contributing/conventions) document.

Reviewer: ```
...and for more details on the review process, refer to our [PR Conventions](../contributing/conventions) document.
```
Let's put this in the `Before opening a Pull Request` section.

Author: We already have it in the "After you open a PR" so removed it from here.

**Note**: if you are a technology partner, make sure you have reviewed the use cases with your [Cortex XSOAR Alliances Team](mailto:soar.alliances@paloaltonetworks.com) and that you have a *Partner ID* to associate your Pull Request to.

A good working example that summarizes all of the above is the [Hello World Content Pack](https://github.com/demisto/content/tree/master/Packs/HelloWorld) that you can use as a reference. Also check out the [Hello World Design Document](https://docs.google.com/document/d/1wETtBEKg37PHNU8tYeB56M1LE314ux086z3HFeF_cX0).

This guide doesn't cover all the topics: please browse the left sidebar and use the search bar to find what you need, and reach out for help over [Slack](https://start.paloaltonetworks.com/join-our-slack-community) on the `#demisto-developers` channel when in doubt.

## Before opening a Pull Request

In order to be able to submit a Pull Request to the Cortex XSOAR [GitHub Repository](https://github.com/demisto/content), you need to:

> - Make sure to check the content contribution [checklist](../contributing/checklist) to make sure you have created everything you need.

Reviewer: This is mentioned too many times. I think it is better as part of the `Guidelines`

Author: This is now mentioned only twice in the document: once in the guidelines and once here. I believe it's important to keep it in both places because users might read the two sections at different times.

Reviewer: This is true but obviously, this applies to several items in these lists. I am just concerned about listing down too many items making this article too tedious and hard to follow. Obviously, one item is not going to change that.

> - Make sure you follow the general [prerequisites](getting-started-guide#before-you-start), satisfy the [requirements](../contributing/contrib-requirements) and set up the [development environment](../concepts/dev-setup) ([tutorial](../tutorials/tut-setup-dev)).

Reviewer: ```suggestion
- Make sure you follow the general [prerequisites](getting-started-guide#before-you-start-developing), satisfy the [requirements](../contributing/contrib-requirements) and set up the [development environment](../concepts/dev-setup) ([tutorial](../tutorials/tut-setup-dev)).
```

Author: done. Good catch!

- Make sure you are working on a GitHub **fork** of the XSOAR content repository, and **create a branch** for your contribution (do **NOT** work on *master*).
- If you're an XSOAR partner, have your `partner-id` (this should have been communicated to you over the onboarding emails from the Alliance team).

> - Design Document: we encourage you to prepare a Design Document that describes the capabilities of your Pack. Usually it's a Google Doc shared with you by the Alliances team. Click [here](https://docs.google.com/document/d/1wETtBEKg37PHNU8tYeB56M1LE314ux086z3HFeF_cX0) to see a compiled example of a Design Document.

Reviewer: This should also be moved to guidelines IMO

Author: done
